Initial access point creation implementation

HipsterBrown opened the pull request:
Fixes https://github.com/tessel/t2-cli/issues/87

This configures a 2nd [`wifi-iface`](https://github.com/tessel/openwrt-tessel/blob/master/files/etc/config/wireless) with the options passed in through the cli, enables it, commits the configuration, and resets the `wifi`, `dnsmasq`, and `udhcpd`. 

TODO:
- [x] add tests
- [x] add command to trigger the access point `on` or `off`

HipsterBrown committed:
feat(ap): initial access point creation implementation

johnnyman727 commented:
@HipsterBrown you're killing it! So excited for this feature.

rwaldron commented on lib/tessel/commands.js:
```diff
@@ -88,3 +88,15 @@ module.exports.sysupgrade = function(path) {
+module.exports.setAccessPointSecurity = function(security) {
+  return ['uci', 'set', 'wireless.@wifi-iface[1].encryption=' + security];
+};
+module.exports.turnAccessPointOn = function(enabled) {
```
`turnAccessPointOn(false)` doesn't make sense ;)

Suggest: `accessPointOn`, `accessPointOff`, with no args

HipsterBrown replied:
I was using the `wifi` commands as a reference for naming, i.e. [`turnOnWifi`](https://github.com/tessel/t2-cli/pull/405/files#diff-ccb4fa979baa5d5d841b51a98779dfc9R55). I do like the clear naming connections to the commands being executed on the OpenWRT platform, so I'm open to changing them for ease of use.

HipsterBrown committed:
fix(ap): now able to configure ap setup

rwaldron commented on lib/tessel/access-point.js:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    if (!password || !security) {
+      if (!security) {
```
Shouldn't this be: 

```js
if (password || security) {
  if (security) {
```

Otherwise:

```
$ t2 ap --ssid=foo
INFO Looking for your Tessel...
INFO Connected to bishop over LAN
INFO Setting Access Point with SSID: foo and password: undefined and security mode: undefined
...
```

But I think you wanted it to display: 

```
$ t2 ap --ssid=foo
INFO Looking for your Tessel...
INFO Connected to bishop over USB
INFO Setting Access Point with SSID: foo
...
```

?

HipsterBrown replied:
Yea, I realized that today. I'll be fixing that tonight. Thanks for catching it as well.

rwaldron commented:
Aside from the one note I left, this looks great so far. Needs tests though—let me know if you need help getting started with that.

HipsterBrown committed:
fix: working access point configuration

HipsterBrown committed:
feat(access-point): adds ability to enable/disable AP

rwaldron commented:
> Aside from the one note I left, this looks great so far. Needs tests though—let me know if you need help getting started with that.

Also, don't forget to run `grunt` locally

HipsterBrown commented:
@rwaldron Yea I forgot about that until I started writing tests last night and saw all the `jshint` errors. Should have tests pushed up tonight or tomorrow. :)

rwaldron commented:
@HipsterBrown awesome, please ping me when you want a set of eyes for review

HipsterBrown committed:
tests(access-point): initial implementation

rwaldron commented on test/unit/access-point.js:
```diff
@@ -0,0 +1,267 @@
+var TesselSimulator = require('../common/tessel-simulator');
+
+exports['Tessel.prototype.createAccessPoint'] = {
+  setUp: function(done) {
```
`this.sandbox = sinon.sandbox.create();`

rwaldron commented on test/unit/access-point.js:
```diff
@@ -0,0 +1,267 @@
+    this.commitWirelessCredentials = sinon.spy(commands, 'commitWirelessCredentials');
+    this.reconnectWifi = sinon.spy(commands, 'reconnectWifi');
+    this.reconnectDnsmasq = sinon.spy(commands, 'reconnectDnsmasq');
+    this.reconnectDhcp = sinon.spy(commands, 'reconnectDhcp');
```
replace `sinon.` with `this.sandbox.`

rwaldron commented on test/unit/access-point.js:
```diff
@@ -0,0 +1,267 @@
+    this.commitWirelessCredentials.restore();
+    this.reconnectWifi.restore();
+    this.reconnectDnsmasq.restore();
+    this.reconnectDhcp.restore();
```
```diff
- this.createAccessPoint.restore();
- this.logsInfo.restore();
- this.setAccessPointSSID.restore();
- this.setAccessPointPassword.restore();
- this.setAccessPointSecurity.restore();
- this.commitWirelessCredentials.restore();
- this.reconnectWifi.restore();
- this.reconnectDnsmasq.restore();
- this.reconnectDhcp.restore();
+ this.sandbox.restore();
```

HipsterBrown replied:
Ah that's so much cleaner! Thanks for the heads up. :+1:

rwaldron commented on lib/controller.js:
```diff
@@ -623,6 +623,27 @@ controller.connectToNetwork = function(opts) {
+    return tessel.disableAccessPoint();
+  });
+};
+
```
I like the straight forward naming you went with, major +9001 there.

rwaldron commented on lib/tessel/commands.js:
```diff
@@ -88,3 +88,24 @@ module.exports.sysupgrade = function(path) {
+};
+module.exports.reconnectDhcp = function() {
+  return ['/etc/init.d/odhcpd', 'restart'];
+};
```
So clear, very awesome

rwaldron commented on test/unit/access-point.js:
```diff
@@ -0,0 +1,267 @@
+    this.commitWirelessCredentials = sinon.spy(commands, 'commitWirelessCredentials');
+    this.reconnectWifi = sinon.spy(commands, 'reconnectWifi');
+    this.reconnectDnsmasq = sinon.spy(commands, 'reconnectDnsmasq');
+    this.reconnectDhcp = sinon.spy(commands, 'reconnectDhcp');
```
Do the same changes I mentioned above to this set of tests.

rwaldron commented on test/unit/access-point.js:
```diff
@@ -0,0 +1,267 @@
+    this.commitWirelessCredentials = sinon.spy(commands, 'commitWirelessCredentials');
+    this.reconnectWifi = sinon.spy(commands, 'reconnectWifi');
+    this.reconnectDnsmasq = sinon.spy(commands, 'reconnectDnsmasq');
+    this.reconnectDhcp = sinon.spy(commands, 'reconnectDhcp');
```
And this set

rwaldron commented:
This looks great! Can you write up a "steps to reality check"? I'm guessing an openwrt update is needed? Need that run down—thanks!

HipsterBrown commented:
@rwaldron Thanks for the review and feedback. I'm going to clean up those tests tonight and try a small refactor on the `commitAndClose` function based on @reconbot's feedback to make it more readable.

Is there an example of a "steps to reality check" that I can reference? I'm not sure what's involved with writing up something like that. 

The `openwrt-tessel` update PR is here: https://github.com/tessel/openwrt-tessel/pull/31

HipsterBrown committed:
refactor(access-point): simpler, cleaner tests and source code

johnnyman727 commented:
I copied over the four modified files from [the OpenWRT pull request](https://github.com/tessel/openwrt-tessel/pull/31) and tried to create an access point. Unfortunately, it seemed to hang and then drop Tessel off my network:

```➜  t2-cli git:(HipsterBrown-ap-command) t2 ap -n Jons-Network -p secret -s wep
INFO Looking for your Tessel...
INFO Connected to Henry over LAN
INFO Setting Access Point with SSID: Jons-Network and password: secret and security mode: wep
```

I'll have to dive a little deeper this weekend.

HipsterBrown commented:
@johnnyman727 that is odd. I'll look into that this afternoon a bit. You're saying it won't connect to your wifi network and it won't emit the network you created? 

I think I need to update the `ap` instructions for allowed security options as well.

johnnyman727 commented:
@HipsterBrown yeah, all the network functionality seemed to have dropped out.

I also just updated my OSX to El Capital and that seems to have made dterm unusable. I'll have to find another way into my Tessel.

rwaldron commented:
> I also just updated my OSX to El Capital and that seems to have made dterm unusable. I'll have to find another way into my Tessel.

I discovered the same issue at the jQuery meeting day when @flaki  and I were working on open network connections.

HipsterBrown commented:
Luckily my work computer is still on Yosemite because the IT team told us to hold off on upgrading, so I can use that for dterm in the meantime. 

@johnnyman727 how did you copy over the `openwrt-tessel` files? And can you confirm they were setup correctly?

johnnyman727 commented:
@HipsterBrown I just copied the contents from your PR and pasted them into the appropriate files. I can't confirm they are correct just yet - I need to figure out this dterm thing.

HipsterBrown commented:
@johnnyman727 I reset my config files to match the openwrt-tessel PR and had some issues as well, especially when trying to set the security as `wep`. I'll tinker with those config files a bit more and try to fix the issue. I didn't have an issue with it dropping off the network though.

johnnyman727 commented:
@HipsterBrown sounds good. Did you happen to keep the file contents when they were working as expected?

HipsterBrown commented:
@johnnyman727 I have it working on my Tessel now with security set as `psk2`. I'm going to try removing the default security setting in the `ap` config of `etc/config/wireless` and test that. I also had some issues with turning on the access point again after using `t2 ap off`.

johnnyman727 commented:
Hmm so maybe it was the `wep` security that killed it for my Tessel. In any case, if you can figure out which security modes do or don't work, it would be great to just only allow those security options for now. psk2 should hit most use cases.

HipsterBrown commented:
Sounds good. I'm so excited to get this shipped now that I've figured out how to run a server on Tessel. :smile:

HipsterBrown committed:
fix(ap/wifi): working implementation

This commit includes better error handling and security settings for the
access point creation method. Also included is setting the default wifi
encryption from the connectToNetwork method instead of having it as a
device default, which is crucial for setting up an access point without
connecting to wifi.

HipsterBrown commented:
Solved the issue @johnnyman727 and I encountered. I figured out that having the wifi encryption option set in the default openwrt-tessel image was the blocking the creation/authentication of an access point.

I have updated my openwrt-tessel PR to reflect this new discovery and added setting the encryption into the `connectToNetwork` method with tests. This should help with https://github.com/tessel/t2-cli/issues/393 (setting wifi security) later on.

johnnyman727 commented:
Nice job @HipsterBrown! 

By the way, #434 should fix the error that AppVeyer is throwing on your branch. Once I merge it, rebase on master and we should be passing all the tests.

HipsterBrown commented:
@johnnyman727 Awesome! Have you had a chance to try out the new config?

johnnyman727 commented:
No, not yet. I tried about 6 different serial emulators and none of them seem to work on OSX El Capitan. I'll have to grab another Tessel.

johnnyman727 commented:
No, not yet. I tried about 6 different serial emulators and none of them seem to work on OSX El Capitan. I'll have to grab another Tessel.

rwaldron committed:
Adds check step to AP setup. Needs AP _new_ iface.

johnnyman727 commented:
@HipsterBrown I grabbed a fresh T2, copied over the contents of https://github.com/tessel/openwrt-tessel/pull/31 with the exception of leaving my current wifi network, ran `wifi` (to reboot the network) and my Tessel couldn't get it's network connection back. Do you have an example of the four files from your PR running on your Tessel and it holding a station connection?

HipsterBrown commented:
@johnnyman727 So @rwaldron and I tried a smoke test of the `ap` setup using the latest `master` branch of tessel/openwrt-tessel but had trouble with the default files not writing to the Tessel. We came up with a solution of checking to see if a second `wifi-iface` exists before trying to set the access point config, and if it doesn't exist, add the `wifi-iface` and other required files for `ap` to work. This will reduce the reliance on default configurations for the openwrt-tessel image.

rwaldron commented:
@HipsterBrown @johnnyman727 confirm!

HipsterBrown committed:
fix(access-point): creates boilerplate for new access point when required

HipsterBrown committed:
replaces new Errors with strings

rwaldron commented:
The openwrt-tessel build that I used can be used by doing the following: 

1. SSH to T2
1. Update: 

  ```
  cd /tmp
  wget https://dl.dropboxusercontent.com/u/3531958/openwrt-ramips-mt7620-tessel-squashfs-sysupgrade.bin
  sysupgrade openwrt-ramips-mt7620-tessel-squashfs-sysupgrade.bin
  ```


Reality check/smoke tests: 


1. Create a new ap

  ```
  t2 ap --name maria --ssid maria-net --pass metropolis --security psk2
  ```
  1. On successful creation, connect local machine to that network
    1. SSH to Tessel

      ```
      ssh root@maria.local -i ~/.tessel/id_rsa
      ```
    1. Confirm connection

      ```
      $ ssh root@maria.local -i ~/.tessel/id_rsa
      (elided RSA authenticity output)


      BusyBox v1.23.2 (2015-11-13 19:03:31 UTC) built-in shell (ash)

      Tessel 2  /  Built on OpenWrt
      root@maria:~#
      ```
    1. Disconnect

      ```
      root@maria:~# exit
      Connection to maria.local closed.
      ```
    1. Connect local machine back to ISP network
  1. On failure, report to PR/issue
1. Disable the ap
  
  ```
  t2 ap off
  ```
  1. Confirm that network is no longer available
  1. On failure, report to PR/issue
1. Enable the ap
  
  ```
  t2 ap on
  ```  
  1. Confirm that network is available
  1. On failure, report to PR/issue
1. Change password for ap 
  
  ```
  t2 ap --name maria --ssid maria-net --pass m3tr0p01i5
  ```
  1. On successful creation, repeat steps 1.i.a, 1.i.b, 1.i.c, and 1.i.d
  1. On failure, report to PR/issue
1. Remove security for ap
  
  ```
  t2 ap --name maria --ssid maria-net
  ```
  1. Confirm that network is available
    1. Confirm that network is no longer secure (requires no password to access)
      1. (Your machine may warn about the security of previous connections, respond in the affirmative to proceed)
    1. Repeat steps 1.i.a, 1.i.b, 1.i.c, and 1.i.d
  1. Repeat steps 2 and 3
  1. On failure, report to PR/issue
1. Restore security for ap
  
  ```
  t2 ap --name maria --ssid maria-net --pass metropolis --security psk2
  ```
  1. Repeat steps 1.i.a, 1.i.b, 1.i.c, and 1.i.d
  1. Repeat steps 2 and 3
  1. On failure, report to PR/issue
1. Run program while connected to ap
  
  ```
  t2 ap --name maria --ssid maria-net --pass metropolis --security psk2
  ```
  1. On successful creation, **connect local machine to that network**
  1. Run program from any valid T2 project. For the sake of being thorough, I walked through the entire process: 

    ```
    rwaldron at new-host-4 in ~/robotz
    $ mkdir tessel-init
    rwaldron at new-host-4 in ~/robotz
    $ cd tessel-init
    rwaldron at new-host-4 in ~/robotz/tessel-init
    $ t2 init
    Initializing tessel repository...
    Created package.json.
    Wrote 'Hello World' to index.
    rwaldron at new-host-4 in ~/robotz/tessel-init
    $ t2 run index.js
    INFO Looking for your Tessel...
    INFO Connected to maria over LAN
    INFO Generating slim build.
    INFO Writing project to RAM on maria (2.048 kB)...
    INFO Deployed.
    INFO Running bundled project...
    I'm blinking! (Press CTRL + C to stop)
    ...
    ```
  1. On failure, report to PR/issue
1. Test all failure conditions: 
  1. Omit ssid

    ```
    t2 ap --name maria --pass metropolis --security psk2
    ```
    1. Expect warning:

      ```
      WARN Invalid credentials. Must set ssid
      ```
    1. If no warning, report to PR/issue
  1. Omit password, but include security

    ```
    t2 ap --name maria --ssid maria-net --security psk2
    ```
    1. Expect warning:

      ```
      WARN Invalid credentials. Must set a password with security option
      ```
    1. If no warning, report to PR/issue
  1. Attempt invalid security option
    
    ```
    t2 ap --name maria --ssid maria-net --pass metropolis --security poop
    ```
    1. Expect warning:

      ```
      WARN poop is not a valid security option. Please choose on of the following: none, wep, psk, psk2, wpa, wpa2
      ```
    1. If no warning, report to PR/issue



Throughout testing...

- Keep an eye on the WLAN light on the board. It should be:
  - **on** when the AP is **disabled**
  - **off** when the AP is **enabled**
- When appropriate, be sure to SSH in and `cat /etc/config/wireless`, which will have interesting information: 

```
root@maria:~# cat /etc/config/wireless

config wifi-device 'radio0'
  option type 'mac80211'
  option channel '11'
  option hwmode '11g'
  option path '10180000.wmac'
  option htmode 'HT20'
  option disabled '0'

config wifi-iface
  option device 'radio0'
  option network 'wifi'
  option mode 'sta'
  option encryption 'psk2'
  option key 'XCCGM3C43SC7JVDX'
  option ssid 'JLY4Q_EXT'

config wifi-iface
  option device 'radio0'
  option network 'lan'
  option mode 'ap'
  option key 'metropolis'
  option disabled '0'
  option ssid 'maria-net'
  option encryption 'none'
  ```

rwaldron commented:
I will have make a PR to this branch with the operation-specific success messages.

rwaldron committed:
Tweak status messages

rwaldron commented:
I've been through my smoke tests list twice now and I haven't found any issues that need to be addressed beyond the message tweaks, but I made a PR for that https://github.com/HipsterBrown/t2-cli/pull/4 

@johnnyman727 this is good to go when you're ready

HipsterBrown committed:
Merge pull request #4 from rwaldron/ap-command

Tweak status messages

HipsterBrown committed:
fix(access-point): removes template literals cause jsbeautify

HipsterBrown commented:
Fixed up the JSBeautify errors. I would love to use template literals @rwaldron but I'm not sure how to override those settings yet. 

This should be ready to merge. I can squash those commits before that if need be.

rwaldron commented:
@HipsterBrown I just updated grunt-jscs to the latest version and the issue is resolved. Don't worry about it for now, we can revisit later

rwaldron commented:
I've just done a complete run through with this branch, on a T2 that was fresh out of the anti-static bag (which is the closest thing to end developer/user that I can think of). This is really great work and I'm going to land it!

HipsterBrown commented:
:raised_hands: :clap: I'll start working on the tutorial and docs tonight.
